qualang_tools/plot/Fitting/fitting_functions.py

```diff
@@ -0,0 +1,348 @@
+from scipy import optimize
+import numpy as np
+
+
+def fit_ramsey(x, y):
+    """
+    Create a fit to Ramsey experiment of the form
+
+    .. math::
+    f(x) = uncertainty_population * (1 - np.exp(-x * (1/tau))) + amp / 2 * (
+        np.exp(-x * (1/tau))
+        * (initial_offset * 2 + np.cos(2 * np.pi * f * x + phase))
+        )
+
+    for unknown parameters `f`, `phase`, `tau`, `amp`, `uncertainty_population` and `initial_offset`.
+
+    :param x:  The data on the x axis
+    :param y: The data on the y axis
+    :return: A dictionary of (fit_func, f, phase, tau, amp, uncertainty_population, initial_offset)
+        fit_func - The fitted function
+        f -  The detuning frequency
+        phase - The phase
+        tau - The decay constant
+        amp - The amplitude
+        uncertainty_population -  The uncertainty in population
+        initial_offset - The initial offset
+    """
+    w = np.fft.fft(y)
+    freqs = np.fft.fftfreq(len(x))
+    new_w = w[1 : len(freqs // 2)]
+    new_f = freqs[1 : len(freqs // 2)]
+
+    ind = new_f > 0
+    new_f = new_f[ind]
+    new_w = new_w[ind]
+
+    yy = np.abs(new_w)
+    first_read_data_ind = np.where(yy[1:] - yy[:-1] > 0)[0][0]  # away from the DC peak
+
+    new_f = new_f[first_read_data_ind:]
+    new_w = new_w[first_read_data_ind:]
+
+    out_freq = new_f[np.argmax(np.abs(new_w))]
+    new_w_arg = new_w[np.argmax(np.abs(new_w))]
+
+    # Finding a guess for the frequency
+    omega = out_freq * 2 * np.pi / (x[1] - x[0])
+
+    cycle = int(np.ceil(1 / out_freq))
+    peaks = (
+        np.array(
+            [np.std(y[i * cycle : (i + 1) * cycle]) for i in range(int(len(y) / cycle))]
+        )
+        * np.sqrt(2)
+        * 2
+    )
+
+    # Finding a guess for the offset
+    initial_offset = np.mean(y[:cycle])
+    cycles_wait = np.where(peaks > peaks[0] * 0.37)[0][-1]
+
+    post_decay_mean = np.mean(y[-cycle:])
+
+    # Finding a guess for the decay
+    decay_guess = (
+        np.log(peaks[0] / peaks[cycles_wait]) / (cycles_wait * cycle) / (x[1] - x[0])
+    )  # get guess for decay #here
+
+    fit_type = lambda x, a: post_decay_mean * a[4] * (
+        1 - np.exp(-x * decay_guess * a[1])
+    ) + peaks[0] / 2 * a[2] * (
+        np.exp(-x * decay_guess * a[1])
+        * (
+            a[5] * initial_offset / peaks[0] * 2
+            + np.cos(2 * np.pi * a[0] * omega / (2 * np.pi) * x + a[3])
+        )
+    )
+
+    def curve_fit(f, x, y, a0):
+        def opt(x, y, a):
+            return np.sum(np.abs(f(x, a) - y) ** 2)
+
+        out = optimize.minimize(lambda a: opt(x, y, a), a0)
```

Comment:
You are right, it's important to check it. In order to add this flag I first have to make a test. Do you have any idea checking it?

Reply:
The variable out.success will automatically be False if the fit does not converge so one way of testing it would be try to fit a Lorentzian with the Ramsey fitting function for instance

Reply:
Oh, ok.. cool thanks !